### PR TITLE
ovl_Eff_Lastday Decompiled

### PR DESCRIPTION
**Pull request by sonicdcer**

### CODE REVIEW

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c [hunk 1]
```diff
@@ -33,25 +33,201 @@ const ActorInit Eff_Lastday_InitVars = {
...
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Eff_Lastday/func_80BEC0A4.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Eff_Lastday/EffLastday_Draw.s")
+void EffLastday_Init(Actor* thisx, PlayState* play2) {
+    PlayState* play = play2;
+    EffLastday* this = THIS;
+
+    Actor_SetScale(&this->actor, 0.1f);
+    switch (this->actor.params & 0xF) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c:41]: make a macro for this

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c [hunk 2]
```diff
@@ -33,25 +33,201 @@ const ActorInit Eff_Lastday_InitVars = {
...
+        case 1:
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c:42]: enum

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c [hunk 3]
```diff
@@ -33,25 +33,201 @@ const ActorInit Eff_Lastday_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Eff_Lastday/func_80BEC0A4.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_Eff_Lastday/EffLastday_Draw.s")
+void EffLastday_Init(Actor* thisx, PlayState* play2) {
+    PlayState* play = play2;
+    EffLastday* this = THIS;
+
+    Actor_SetScale(&this->actor, 0.1f);
+    switch (EFFLASTDAY_GET_F(&this->actor)) {
+        case ACTION_1:
```
Comment (hensldm) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c:42]: These case's are not actions. they should use the params enum.

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c [hunk 4]
```diff
@@ -33,25 +33,201 @@ const ActorInit Eff_Lastday_InitVars = {
...
+            Actor_SetScale(&this->actor, 0.5f);
+            break;
+        case ACTION_3:
+            this->dList = object_lastday_DL_000060;
+            this->matAnim = Lib_SegmentedToVirtual(object_lastday_Matanimheader_000148);
+            this->actionCmd = 0x1FE;
+            this->actionFunc = func_80BEBF78;
+            Actor_SetScale(&this->actor, 0.2f);
+            this->actor.home.rot.z = 0;
+            break;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c:63]:
```suggestion
            break;
            
        case ACTION_2:
            this->dList = object_lastday_DL_000210;
            this->matAnim = Lib_SegmentedToVirtual(object_lastday_Matanimheader_000308);
            this->actionCmd = 0x1FD;
            this->actionFunc = func_80BEBEB8;
            Actor_SetScale(&this->actor, 0.5f);
            break;
            
        case ACTION_3:
            this->dList = object_lastday_DL_000060;
            this->matAnim = Lib_SegmentedToVirtual(object_lastday_Matanimheader_000148);
            this->actionCmd = 0x1FE;
            this->actionFunc = func_80BEBF78;
            Actor_SetScale(&this->actor, 0.2f);
            this->actor.home.rot.z = 0;
            break;
            
```

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c [hunk 5]
```diff
@@ -33,25 +33,201 @@ const ActorInit Eff_Lastday_InitVars = {
...
+                    this->alpha = this->alpha + 10;
+                }
+                if (this->alpha > 255) {
+                    this->alpha = 255;
+                }
+                this->step++;
+                break;
+            case ACTION_3:
+                this->actor.draw = EffLastday_Draw;
+                break;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c:108]:
```suggestion
                break;
                
            case ACTION_2:
                this->actor.draw = NULL;
                this->actor.draw = EffLastday_Draw;
                if (this->alpha < 255) {
                    this->alpha = this->alpha + 10;
                }
                if (this->alpha > 255) {
                    this->alpha = 255;
                }
                this->step++;
                break;
                
            case ACTION_3:
                this->actor.draw = EffLastday_Draw;
                break;
```

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c [hunk 6]
```diff
@@ -33,25 +33,201 @@ const ActorInit Eff_Lastday_InitVars = {
...
+                this->actor.draw = NULL;
+                this->step = 0;
+                break;
+            case ACTION_2:
+                this->actor.draw = EffLastday_Draw;
+                this->step++;
+                break;
+            case ACTION_3:
+                this->actor.draw = EffLastday_Draw;
+                break;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c:132]:
```suggestion
                break;
                
            case ACTION_2:
                this->actor.draw = EffLastday_Draw;
                this->step++;
                break;
                
            case ACTION_3:
                this->actor.draw = EffLastday_Draw;
                break;
```

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c [hunk 7]
```diff
@@ -33,25 +33,201 @@ const ActorInit Eff_Lastday_InitVars = {
...
```
Comment (hensldm) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c:156]:
```suggestion
                break;
                
            case ACTION_2:
                this->actor.draw = EffLastday_Draw;
                this->step++;
                break;
                
            case ACTION_3:
                this->actor.draw = EffLastday_Draw;
                break;
```

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c [hunk 8]
```diff
@@ -33,25 +33,201 @@ const ActorInit Eff_Lastday_InitVars = {
...
+                    this->alpha += 5;
+                }
+                if (this->alpha > 255) {
+                    this->alpha = 255;
+                }
+                this->step++;
+                break;
+            case ACTION_3:
+                this->actor.draw = EffLastday_Draw;
+                break;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c:191]:
```suggestion
                break;
                
            case ACTION_2:
                if (!this->actor.home.rot.z) {
                    this->actor.home.rot.z = true;
                    Actor_PlaySfxAtPos(&this->actor, NA_SE_EV_COMING_FIRE);
                }
                this->actor.draw = EffLastday_Draw;
                if (this->alpha < 255) {
                    this->alpha += 5;
                }
                if (this->alpha > 255) {
                    this->alpha = 255;
                }
                this->step++;
                break;
                
            case ACTION_3:
                this->actor.draw = EffLastday_Draw;
                break;
```

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c [hunk 9]
```diff
@@ -33,25 +33,201 @@ const ActorInit Eff_Lastday_InitVars = {
...
+            default:
+                this->actor.draw = NULL;
+                this->alpha = 0;
+                this->step = 0;
+                break;
+            case ACTION_2:
+                this->actor.draw = NULL;
+                this->actor.draw = EffLastday_Draw;
+                if (this->alpha < 255) {
+                    this->alpha = this->alpha + 10;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c:99]:
```suggestion
                    this->alpha += 10;
```

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c [hunk 10]
```diff
@@ -33,25 +33,201 @@ const ActorInit Eff_Lastday_InitVars = {
...
+    }
+}
+
+void EffLastday_Update(Actor* thisx, PlayState* play) {
+    EffLastday* this = THIS;
+
+    this->actionFunc(this, play);
+}
+
+void EffLastday_Draw_SetVtxAlpha(s16 alpha) {
```
Comment (hensldm) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c:204]:
```suggestion
void EffLastday_SetVtxAlpha(s16 alpha) {
```

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.h [hunk 11]
```diff
@@ -3,16 +3,36 @@
...
+    /* 0x144 */ Gfx* dList;
+    /* 0x148 */ AnimatedMaterial* matAnim;
+    /* 0x14C */ s16 step;
+    /* 0x14E */ s16 alpha;
+    /* 0x150 */ s16 actionCmd;
+    /* 0x154 */ EffLastdayActionFunc actionFunc;
 } EffLastday; // size = 0x158
 
+typedef enum EffLastDayAction {
+    /* 0x0 */ ACTION_NONE,
```
Comment (hensldm) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.h:23]:
```suggestion
    /* 0x0 */ ACTION_0,
```
Personally I don't think you should use `NONE` here (same for Params below) as `NONE` implies that no action is going on (which might be happening here, but it is hard to tell without more documentation. I think just using 0 is fine for now.

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c [hunk 12]
```diff
@@ -33,25 +33,213 @@ const ActorInit Eff_Lastday_InitVars = {
...
+            this->matAnim = Lib_SegmentedToVirtual(object_lastday_Matanimheader_000448);
+            this->actionCmd = 0x1FB;
+            this->actionFunc = func_80BEBD0C;
+            Actor_SetScale(&this->actor, 1.0f);
+            Actor_SpawnAsChild(&play->actorCtx, &this->actor, play, ACTOR_EFF_LASTDAY, this->actor.world.pos.x,
+                               this->actor.world.pos.y, this->actor.world.pos.z, 0, this->actor.world.rot.y, 0, 1);
+            Actor_SpawnAsChild(&play->actorCtx, &this->actor, play, ACTOR_EFF_LASTDAY, this->actor.world.pos.x,
+                               this->actor.world.pos.y, this->actor.world.pos.z, 0, this->actor.world.rot.y, 0, 2);
+            Actor_SpawnAsChild(&play->actorCtx, &this->actor, play, ACTOR_EFF_LASTDAY, this->actor.world.pos.x,
+                               this->actor.world.pos.y, this->actor.world.pos.z, 0, this->actor.world.rot.y, 0, 3);
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c:78]: These should all use the params enum

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c [hunk 13]
```diff
@@ -33,25 +33,213 @@ const ActorInit Eff_Lastday_InitVars = {
...
+            break;
+    }
+}
+
+void EffLastday_Destroy(Actor* thisx, PlayState* play) {
+}
+
+void func_80BEBD0C(EffLastday* this, PlayState* play) {
+    u16 action;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c:87]: Same as usual, `csAction`.

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c [hunk 14]
```diff
@@ -33,25 +33,213 @@ const ActorInit Eff_Lastday_InitVars = {
...
+                this->actor.draw = EffLastday_Draw;
+                break;
+        }
+    } else {
+        this->actor.draw = NULL;
+    }
+}
+
+void func_80BEBEB8(EffLastday* this, PlayState* play) {
+    u16 action;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.c:147]: And here.

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.h [hunk 15]
```diff
@@ -3,16 +3,36 @@
...
+    /* 0x150 */ s16 actionCmd;
     /* 0x154 */ EffLastdayActionFunc actionFunc;
 } EffLastday; // size = 0x158
 
+typedef enum EffLastDayAction {
+    /* 0x0 */ EFFLASTDAY_ACTION_NONE,
+    /* 0x1 */ EFFLASTDAY_ACTION_1,
+    /* 0x2 */ EFFLASTDAY_ACTION_2,
+    /* 0x3 */ EFFLASTDAY_ACTION_3
+} EffLastDayAction;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.h:27]: This is only going to be used in this actor, so can go in the C file.

File: src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.h [hunk 16]
```diff
@@ -3,16 +3,36 @@
...
 typedef void (*EffLastdayActionFunc)(struct EffLastday*, PlayState*);
 
 typedef struct EffLastday {
     /* 0x000 */ Actor actor;
-    /* 0x144 */ char unk_144[0x10];
+    /* 0x144 */ Gfx* dList;
+    /* 0x148 */ AnimatedMaterial* matAnim;
+    /* 0x14C */ s16 step;
+    /* 0x14E */ s16 alpha;
+    /* 0x150 */ s16 actionCmd;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_Eff_Lastday/z_eff_lastday.h:18]: Would use csActionCmd for this too.

**Review by AngheloAlf — APPROVED** [review]

**Review by hensldm — APPROVED** [review]

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**